File: Bai_8/main.py
while False: # Cách 1
   # Đọc dữ liệu từ file input
    with open('MUAHOA.INP', 'r') as file:
        n, m = map(int, file.readline().split())
        flowers = [list(map(int, file.readline().split())) for _ in range(n)]

    # Sắp xếp loại hoa theo giá từ thấp đến cao
    flowers.sort(key=lambda x: x[0])

    total_flowers = 0
    total_cost = 0

    # Duyệt qua từng loại hoa
    for flower in flowers:
        price, quantity = flower
        if total_cost + price * quantity <= m:
            total_cost += price * quantity
            total_flowers += quantity
        else:
            can_buy = (m - total_cost) // price  # Số lượng hoa có thể mua với số tiền còn lại
            total_flowers += can_buy
            break

    # Ghi kết quả vào file output
    with open('MUAHOA.OUT', 'w') as file:
        file.write(str(total_flowers))

while True: # Cách 2
    # Đọc dữ liệu từ file input
    with open('MUAHOA.INP', 'r') as file:
        n, m = map(int, file.readline().split())
        flowers = [list(map(int, file.readline().split())) for _ in range(n)]

    # Sắp xếp loại hoa theo giá từ thấp đến cao
    flowers.sort(key=lambda x: x[0])

    total_flowers = 0
    total_cost = 0

    # Duyệt qua từng loại hoa
    for price, quantity in flowers:
        # Số lượng hoa có thể mua với số tiền hiện tại và số lượng hoa còn lại trong cửa hàng
        can_buy = min(quantity, (m - total_cost) // price)
        total_cost += can_buy * price  # Cập nhật tổng số tiền đã sử dụng
        total_flowers += can_buy  # Cập nhật tổng số lượng hoa đã mua

    # Ghi kết quả vào file output
    with open('MUAHOA.OUT', 'w') as file:
        file.write(str(total_flowers))

File: Bai_8/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

int main()
{
    ifstream file("MUAHOA.INP");
    int n, m;
    file >> n >> m;
    vector<vector<int>> flowers(n, vector<int>(2));

    for (int i = 0; i < n; ++i)
    {
        file >> flowers[i][0] >> flowers[i][1];
    }
    file.close();

    sort(flowers.begin(), flowers.end(), [](const vector<int> &a, const vector<int> &b)
         { return a[0] < b[0]; });

    int total_flowers = 0;
    int total_cost = 0;

    for (auto &flower : flowers)
    {
        int price = flower[0];
        int quantity = flower[1];
        int can_buy = min(quantity, (m - total_cost) / price);
        total_cost += can_buy * price;
        total_flowers += can_buy;
    }

    ofstream outfile("MUAHOA.OUT");
    outfile << total_flowers;
    outfile.close();

    return 0;
}
